src/Serializing/SExpression/Precise.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE DataKinds, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, OverloadedStrings, ScopedTypeVariables, TypeApplications, TypeFamilies, TypeOperators, UndecidableInstances #-}
+module Serializing.SExpression.Precise
+( serializeSExpression
+) where
+
+import Data.ByteString.Builder
+import Data.Foldable (fold)
+import Data.List (intersperse)
+import Data.Text (Text)
+import GHC.Generics
+
+serializeSExpression :: ToSExpression t => t -> Builder
+serializeSExpression t = toSExpression t 0 <> "\n"
+
+
+nl :: Int -> Builder
+nl n | n <= 0    = ""
+     | otherwise = "\n"
+
+pad :: Int -> Builder
+pad n = stringUtf8 (replicate (2 * n) ' ')
+
+
+class ToSExpression t where
+  toSExpression :: t -> Int -> Builder
```

Comment:
This might actually be a good case for `Uniplate` or similar, but as I don’t have experience with the syb family of approaches I’ve opted instead to implement it using `GHC.Generics`.

Reply:
TIL about `Uniplate` (well, yesterday, YIL?). Not familiar with advantages over using that to reduce boilerplate vs `GHC.Generics` in this case!

Reply:
Me neither 😅

Reply:
Uniplate would probably make this code a lot shorter, but it would also be slower, rely on RTTI, and would necessitate some shenanigans to handle the `Text` case. We should start getting familiar with Uniplate just so we don’t have to write an advanced overlap every single time we need a generic function, but since this is here and working I’m a big 👍